File: py/Roman_to_Integer.py
class Solution:
    def romanToInt(self, s: str) -> int:
        roman = {'I': 1, 'V': 5, 'X': 10,
                 'L': 50, 'C': 100, 'D': 500, 'M': 1000}
        prev, total = 0, 0
        # for c in s:
        #     curr = roman[c]
        #     total += curr
        #     if curr > prev:
        #         total -= 2 * prev
        #     prev = curr

        for c in s:
            curr = roman[c]
            if curr > prev:
                total = total - prev + (curr - prev)
            else:
                total += curr
            prev = curr
        return total

if __name__ == '__main__':
    s = Solution()
    print(s.romanToInt('IX'))

File: c_plus/Roman_to_Integer.cpp
#include <iostream>
#include <string>

using namespace std;

class Solution {
public:

    int romanCharToInt(char ch){
    int d = 0;
    switch(ch){
        case 'I':  
            d = 1;  
            break;  
        case 'V':  
            d = 5;  
            break;  
        case 'X':  
            d = 10;  
            break;  
        case 'L':  
            d = 50;  
            break;  
        case 'C':  
            d = 100;  
            break;  
        case 'D':  
            d = 500;  
            break;  
        case 'M':  
            d = 1000;  
            break;  
    }
    return d;
};

    int romanToInt(string s) {
        if (s.size() <= 0) return 0;
        int res = romanCharToInt(s[0]);
        for (int i = 1; i < s.size(); i++){
            int prev = romanCharToInt(s[i - 1]);
            int curr = romanCharToInt(s[i]);
            if (prev < curr){
                res = res - prev + (curr - prev);
            }else{
                res += curr;
            }
        }
        return res;
    }
};

int main(){
    Solution s;
    cout << s.romanToInt("IX") << endl;
}